File: graylog2-server/src/main/java/org/graylog/plugins/views/search/QueryMetadata.java
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
package org.graylog.plugins.views.search;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableSet;
import org.graylog.plugins.views.search.elasticsearch.QueryParam;

import java.util.Set;

import static com.google.common.collect.ImmutableSet.of;

@AutoValue
public abstract class QueryMetadata {

    @JsonProperty("used_parameters_names")
    public ImmutableSet<String> usedParameterNames() {
        return usedParameters().stream().map(QueryParam::name).collect(ImmutableSet.toImmutableSet());
    }

    @JsonIgnore
    public abstract ImmutableSet<QueryParam> usedParameters();

    public static QueryMetadata empty() {
        return QueryMetadata.builder().build();
    }

    public static Builder builder() {
        return new AutoValue_QueryMetadata.Builder()
                .usedParameters(of());
    }

    public abstract Builder toBuilder();

    @AutoValue.Builder
    public abstract static class Builder {

        public abstract Builder usedParameters(Set<QueryParam> usedParameters);

        public abstract QueryMetadata build();
    }
}

File: graylog2-web-interface/src/views/logic/search/QueryMetadata.js
/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
// @flow strict

import * as Immutable from 'immutable';

type State = {
  usedParameterNames: Immutable.Set<string>,
};

export type QueryMetadataJson = {
  used_parameters_names: Array<string>,
};

export default class QueryMetadata {
  _value: State;

  constructor(usedParameterNames: Immutable.Set<string>) {
    this._value = { usedParameterNames };
  }

  get usedParameterNames(): Immutable.Set<string> {
    return this._value.usedParameterNames;
  }

  static fromJSON(value: QueryMetadataJson) {
    return new QueryMetadata(Immutable.Set(value.used_parameters_names));
  }
}
